crates/ruff_db/src/system/path.rs

```diff
@@ -306,6 +306,86 @@ impl SystemPath {
     pub fn from_std_path(path: &Path) -> Option<&SystemPath> {
         Some(SystemPath::new(Utf8Path::from_path(path)?))
     }
+
+    /// Makes a path absolute and normalizes it without accessing the file system.
+    ///
+    /// Adapted from [cargo](https://github.com/rust-lang/cargo/blob/fede83ccf973457de319ba6fa0e36ead454d2e20/src/cargo/util/paths.rs#L61)
+    ///
+    /// # Examples
+    ///
+    /// ## Posix paths
+    ///
+    /// ```
+    /// # #[cfg(unix)]
+    /// # fn main() {
+    ///   use ruff_db::system::{SystemPath, SystemPathBuf};
+    ///
+    ///   // Relative to absolute
+    ///   let absolute = SystemPath::absolute("foo/./bar", "/tmp");
+    ///   assert_eq!(absolute, SystemPathBuf::from("/tmp/foo/bar"));
+    ///
+    ///   // Absolute to absolute
+    ///   let absolute = SystemPath::absolute("/foo//test/.././bar.rs", "/tmp");
+    ///   assert_eq!(absolute, SystemPathBuf::from("/foo/bar.rs"));
+    /// # }
+    /// # #[cfg(not(unix))]
+    /// # fn main() {}
+    /// ```
+    ///
+    /// ## Windows paths
+    ///
+    /// ```
+    /// # #[cfg(windows)]
+    /// # fn main() {
+    ///   use ruff_db::system::{SystemPath, SystemPathBuf};
+    ///
+    ///   // Relative to absolute
+    ///   let absolute = SystemPath::absolute(r"foo\.\bar", r"C:\tmp");
+    ///   assert_eq!(absolute, SystemPathBuf::from(r"C:\tmp\foo\bar"));
+    ///
+    ///   // Absolute to absolute
+    ///   let absolute = SystemPath::absolute(r"C:\foo//test\..\./bar.rs", r"C:\tmp");
+    ///   assert_eq!(absolute, SystemPathBuf::from(r"C:\foo\bar.rs"));
+    /// # }
+    /// # #[cfg(not(windows))]
+    /// # fn main() {}
+    /// ```
+    pub fn absolute(path: impl AsRef<SystemPath>, cwd: impl AsRef<SystemPath>) -> SystemPathBuf {
+        fn absolute(path: &SystemPath, cwd: &SystemPath) -> SystemPathBuf {
+            let path = &path.0;
+
+            let mut components = path.components().peekable();
+            let mut ret = if let Some(
+                c @ (camino::Utf8Component::Prefix(..) | camino::Utf8Component::RootDir),
+            ) = components.peek().cloned()
+            {
+                components.next();
+                Utf8PathBuf::from(c.as_str())
+            } else {
+                cwd.0.to_path_buf()
+            };
+
+            for component in components {
+                match component {
+                    camino::Utf8Component::Prefix(..) => unreachable!(),
+                    camino::Utf8Component::RootDir => {
+                        ret.push(component);
+                    }
+                    camino::Utf8Component::CurDir => {}
+                    camino::Utf8Component::ParentDir => {
+                        ret.pop();
+                    }
```

Comment:
Is it worth adding a comment similar to https://github.com/astral-sh/ruff/blob/757c75752e1a728669bb97a412ddf205d165b0ea/crates/ruff_db/src/vendored.rs#L260-L265 here?

Reply:
I added an example test case that also verifies that it works that way.